dss/stepfunctions/s3copyclient/implementation.py

```diff
@@ -0,0 +1,257 @@
+import binascii
+import collections
+import concurrent.futures as futures
+import copy
+
+import hashlib
+import threading
+import typing
+
+import boto3
+from cloud_blobstore.s3 import S3BlobStore
+
+from ...stepfunctions import generator
+from ...stepfunctions.lambdaexecutor import TimedThread
+from ...util.aws import get_s3_chunk_size
+
+
+# CONSTANTS
+LAMBDA_PARALLELIZATION_FACTOR = 32
+CONCURRENT_REQUESTS = 8
+
+
+# Public input/output keys for the state object.
+class Key:
+    SOURCE_BUCKET = "srcbucket"
+    SOURCE_KEY = "srckey"
+    DESTINATION_BUCKET = "dstbucket"
+    DESTINATION_KEY = "dstkey"
+    FINISHED = "finished"
+
+
+# Internal key for the state object.
+class _Key:
+    SOURCE_ETAG = "srcetag"
+    UPLOAD_ID = "uploadid"
+    SIZE = "size"
+    PART_SIZE = "partsz"
+    NEXT_PART = "next"
+    LAST_PART = "last"
+    PART_COUNT = "count"
+
+
+def setup_copy_task(event, lambda_context):
+    source_bucket = event[Key.SOURCE_BUCKET]
+    source_key = event[Key.SOURCE_KEY]
+    destination_bucket = event[Key.DESTINATION_BUCKET]
+    destination_key = event[Key.DESTINATION_KEY]
+
+    s3_blobstore = S3BlobStore()
+    blobinfo = s3_blobstore.get_all_metadata(source_bucket, source_key)
+    source_etag = blobinfo['ETag'].strip("\"")  # the ETag is returned with an extra set of quotes.
+    source_size = blobinfo['ContentLength']  # type: int
+    part_size = get_s3_chunk_size(source_size)
+    part_count = source_size // part_size
+    if part_count * part_size < source_size:
+        part_count += 1
+    if part_count > 1:
+        mpu = s3_blobstore.s3_client.create_multipart_upload(Bucket=destination_bucket, Key=destination_key)
+        upload_id = mpu['UploadId']
+    else:
+        upload_id = None
+
+    event[_Key.SOURCE_ETAG] = source_etag
+    event[_Key.UPLOAD_ID] = upload_id
+    event[_Key.SIZE] = source_size
+    event[_Key.PART_SIZE] = part_size
+    event[_Key.PART_COUNT] = part_count
+    event[Key.FINISHED] = False
+
+    return event
+
+
+def copy_worker(event, lambda_context, branch_id):
+    class CopyWorkerTimedThread(TimedThread[dict]):
```

Comment:
This class does not depend on the parent's args, consider bringing it into the top scope?

Reply:
Is there precedence in Python for using a nested class/function for code organization?  The reason I did this was because I thought `CopyWorkerTimedThread` was never going to be used anywhere else except by `copy_worker`.

Reply:
I guess that's fine, I think there is a negligible performance impact to defining the class at function runtime (unless it makes an I/O call while being defined). Also the fewer indents, the better. But I'm fine keeping it where it is.